_posts/2021-03-05-automatically-generated-api-tests-with-clojure-and-reitit.md

```diff
@@ -0,0 +1,332 @@
+with a valid token, while `with-resource` provides a context with a pre-created
+resource.
+
+```clojure
+(def next-id (atom 0))
+
+(defn- create-token [ctx]
+  (assoc ctx
+    :token (swap! next-id inc)))
+
+(defn- create-resource [ctx]
+  (assoc ctx
+    :resource-id
+    (->
+      (request :post (:token ctx) "/api/resource" {:content "some content"})
+      :body
+      :resource-id)))
+
+(def empty-ctx (constantly {}))
+
+(defn with-token []
+  (-> (empty-ctx)
+      (create-token)))
+
+(defn with-resource []
+  (-> (empty-ctx)
+      (create-token)
+      (create-resource)))
+```
+
+What does `test-path` look like? Based on our usage, it needs to take in one or
+two arguments, the first of which is some function which creates the resources
+needed for testing, and the second is a body which will be sent to the API if
+provided. The function should also replace possible path parameters with
+appropriate id's from the context. In our use case, the only possible path
+parameter is `:resource-id`, but this function is easily extendable for
+additional parameters.
+
+```clojure
+(defn- create-request-path [path ctx]
+  (cond-> path
+          (:resource-id ctx) (string/replace ":resource-id" (-> ctx :resource-id str))))
+
+(defn test-path
+  ([context-fn] (test-path context-fn nil))
+  ([context-fn body]
+   (fn [path method]
+     (let [ctx          (context-fn)
+           request-path (create-request-path path ctx)
+           request-body (if (and (ifn? body) (not (map? body)))
+                          (body ctx)
+                          body)]
+       (testing "Request without tokens returns 403"
+         (is (= {:status 400 :body {:cause :missing-authentication}}
+                (request method nil request-path request-body))))
+       (testing "Request with tokens returns 200"
+         (is (= 200
+                (:status
+                  (request method (:token ctx) request-path request-body))))))
+     ; todo: add a test here which adds / removes the token from the resource
+   )))
+```
+
+With those two pieces of code, the  `test-all-paths` method goes over every
+path in the test-routes object, and creates API requests for the route.
+
+Collecting a list of routes
+---------------------------
+
+If we don't add all of the routes to the test-routes object, nothing in the
+system notifies us about it. Since we can just ask `reitit` for a list of
+routes, let's also test that our test blob matches the routes in the system.
+
+When your API looks something like this:
+```clojure
+(def api
+  [["/api"
+    ["/authorization"
+     {:post   add-authorization
+      :delete remove-authorization}]
+    ["/resource"
+     ["" {:post create-resource}]
+     ["/:resource-id"
+      {:get    read-resource
+       :put    update-resource
+       :delete delete-resource}]]]])
+```
+
+You can use reitit-impl to get a list of routes, and check that all of those are tested:
```

Comment:
Perhaps "the `reitit.impl/resolve-routes` function"? If you just say "reitit-impl", it's not really clear (yet) what it refers to.

`impl` namespaces aren't usually part of the public API of Clojure libs. Is `reitit.impl`?

Reply:
Well, it's probably not public, but it works for this kind of code well enough. I don't think there's an actual public API for that kind of function.